### ISSUE TITLE
Modelgen generates non-codeable model

### ISSUE THREAD

**Issue by MyJBMe** [aws-amplify/amplify-ios#257]
Based on this GraphQL-schema

```
type Todo @model @auth(rules: [{allow: owner}, { allow: owner, ownerField: "inspectors", operations: [read] }]) {
  id: ID!
  name: String!
  description: String
  owner: String
  inspectors: [String]
}
```

modelgen generates following model which does not conform to both encodable und decodable protocols, as the protocol can not handle `List<String>?`.

```
import Amplify
import Foundation

public struct Todo: Model {
  public let id: String
  public var name: String
  public var description: String?
  public var owner: String?
  public var inspectors: List<String>?
  
  public init(id: String = UUID().uuidString,
      name: String,
      description: String? = nil,
      owner: String? = nil,
      inspectors: List<String>? = []) {
      self.id = id
      self.name = name
      self.description = description
      self.owner = owner
      self.inspectors = inspectors
  }
}
```

I fixed this for this model by changing the inspectors type to `[String]?`.

**Comment by rodrigoelp** [comment]
Hi @MyJBMe, when modelgen runs it tends to generate a `Todo.swift` file and a `Todo+Schema.swift` file which contains the extra definition you are looking for. The schema file has an extension applied to the model which implements all the coding keys and what not you need it to work. Make sure you are including those files as well.

As to the `List` model, that is a type introduced by the amplify framework which I think tracks specific things required for it to work.

**Comment by MyJBMe** [comment]
Hi @rodrigoelp, well I understand what it is for, but the problem is, that the generation of this model produces errors. Xcode does not compile a List<String> since this is not conform to the decodable and encodable protocols. It starts working with [String] instead of List<String>, so it would make sense to update the modelgen to generate these as a standard array, wouldn't it?

**Comment by rodrigoelp** [comment]
AH! I can see the problem the code is generating here (I think I misunderstood what you were trying to say).

The intrinsic problem here is the fact that `List<M>` is expecting `M` to be a model, `String` (or any primitive type as the matter of fact) won't follow/implement the restrictions applied to model. That's the reason why changing it to an array fixes the issue.

**Comment by MyJBMe** [comment]
You are right, that one is on me!

So the solution here would be to either generate arrays for primitive types and lists for models or to use `[M]` in general?

**Comment by rodrigoelp** [comment]
The particular case you have is tricky (based on what amplify supports right now)...

If I had to solve this problem, I would pull the `inspectors` field as a model and link it as a parent-child relationship. The problem is, you want to use is as part of `@auth` and I am too new to this framework to understand the proper way to solve it.
Your fix seems decent enough (at least to me) so I would propose the change from `List<M>` to just `[T]` if the `M` doesn't inherit/implements any of the models.

I guess we will need to wait for the AWS guys to comment in here :)

**Comment by drochetti** [comment]
Thanks for the report and for the thoughtful discussion about the issue.

> Your fix seems decent enough (at least to me) so I would propose the change from List<M> to just [T] if the M doesn't inherit/implements any of the models.

Yeah, I believe this is the right approach and we will implement it. Just for context: behind the scenes DataStore uses a SQL database, so we have to do some custom (de)serialization for more complex types and `[String]` is one of the cases.

**Comment by siddharthsharma94** [comment]
Im still having the issue, what exactly is the fix? Making the model into a String[]? Doesn’t seem to fix it.

**Comment by rodrigoelp** [comment]
I wouldn't call it a fix, it was said it is a workaround/awful hack we are forced to do. Also, what error/issue is it reporting on your end?

I created a sample with a similar schema as defined by @MyJBMe  and the updated generated code and it works _fine_. There is no build error saying `String` doesn't conform with `Model` type and there is no runtime error deserialising the type either.

**Comment by drochetti** [comment]
Hey @MyJBMe we just released a stable [v1.0.0](https://github.com/aws-amplify/amplify-ios/releases/tag/v1.0.0). Could you try following the [new DataStore tutorial](https://docs.amplify.aws/start/getting-started/native_setup/q/integration/ios) and see if the new flow and stable code fixes your issues?

We did add support for the scenario you described (`inspectors: [String]`) and that should be supported in `1.0.0`. Can you give it a try and let me know if that fixes the issue?

---

I appreciate your patience and the fact you tried our unstable preview and provided valuable feedback and insights with your issue report. I hope the stable release fixes things for you.

We also have a brand new Discord server for Amplify related topics. Feel free to hit me up there to talk about your use cases and challenges with the new APIs and docs. https://discord.gg/Px5YVfQ

**Comment by drochetti** [comment]
Closing this issue. Feel free to open a new one if you have problems with the stable version of Amplify.
